torch/_dynamo/graph_region_tracker.py

```diff
@@ -0,0 +1,290 @@
+        duplicates = self.loc_to_duplicates[
+            self._hash_node(
+                tx.f_code.co_filename, tx.lineno, tx.instruction_pointer, node
+            )
+        ]
+        duplicates.append(node)
+        self.node_to_duplicates[node] = duplicates
+
+    def is_identical(self, n0: Node, n1: Node) -> bool:
+        return (
+            n0 in self.node_to_duplicates
+            and n1 in self.node_to_duplicates
+            and self.node_to_duplicates[n0] == self.node_to_duplicates[n1]
+            and n0 is not n1
+        )
+
+    def get_identical_regions(self, graph: torch.fx.Graph) -> List[List[Region]]:
+        topological_ranking = {node: i for i, node in enumerate(graph.nodes)}
+        group_ranking = {}
+        region_groups = []
+
+        # Create region groups; a region group is a group
+        # of regions that are all identical. In this initial state
+        # each region in the group is a single node, and we discard
+        # groups that are only a single region.
+        # We track the topological ranking to start with groups later in the graph
+        # the reason for this is that we will necessarily create the largest groups first.
+        for group in self.loc_to_duplicates.values():
+            if len(group) > 1:
+                region_group = []
+                min_rank = math.inf
+                for node in group:
+                    min_rank = min(min_rank, topological_ranking[node])
+                    region_group.append([node])
+
+                region_groups.append(region_group)
+                group_ranking[id(region_group)] = min_rank
+
+        region_groups.sort(key=lambda g: -group_ranking[id(g)])
+
+        # We start from regions later in the graph and expand them earlier
+        # as a result, we will create the largest regions first and they won't
+        # overlap.
+        seen_nodes: Set[Node] = set()
+        for region_group in region_groups:
+            fully_expand_region_group(region_group, seen_nodes, self.is_identical)
+
+        return [
+            region_group for region_group in region_groups if len(region_group[0]) > 1
+        ]
+
+    def __str__(self) -> str:
+        return f"GraphRegionTracker(loc_to_duplicates={self.loc_to_duplicates}, node_to_duplicates={self.node_to_duplicates})"
+
+
+def fully_expand_region_group(
+    regions: List[Region],
+    seen_nodes: Set[Node],
+    is_identical_fn: Callable[[Node, Node], bool],
+) -> None:
+    # All regions should start with 1 node
+    assert all(len(region) == 1 for region in regions)
+    region_iters = []
+    for region in regions:
+        (origin,) = region  # Only works for 1 element sets
+        region_iters.append(BfsRegionIter.create(origin))
+
+    nodes_to_add: List[Node] = []
+
+    # we already have the origin node in each region
+    for region_it in region_iters:
+        _, node = region_it.next()
+        assert node
+        region_it.add_children(node)
```

Comment:
I'm lost, why are we getting the first child of each origin and then adding their children? Are we assuming each node only has one child?

Reply:
so each region iter before this loop has a single node in it. We know all of the initial single nodes in a region group are equivalent (or else their hash would not have been the same), so we can just skip the checks below and start with their children instead. We add all of their children to the iters and start processing.